JEI crafting integration, redux

uecasm opened the pull request:
Closes #6540
Probably closes ldtteam/minecolonies-features#282

# Changes proposed in this pull request:
- Stage one of adding crafting modules to buildings (more work to follow)
- Adds JEI recipe integration

Review please

1. The recipe lookup allows you to see what crafting (or furnace smelting, as appropriate) recipes can be taught to buildings, plus those that they already know via crafterrecipes:
![image](https://user-images.githubusercontent.com/539951/119513926-c0c24c80-bdc8-11eb-9d7a-39094dc48d16.png)
    The latter will also indicate if there are level or research restrictions.
2. There's also a slightly different style used for loot-table-based recipes (sifter and enchanter).  These are determined from the recipe itself, so they can be freely mixed in a single building (e.g. enchanter):
![image](https://user-images.githubusercontent.com/539951/119514265-f9fabc80-bdc8-11eb-8a25-2b2498c652b9.png)
3. Additionally, clicking the arrow in the "teach recipe" UI will take you to the corresponding page that shows learnable recipes.
![image](https://user-images.githubusercontent.com/539951/119514391-17c82180-bdc9-11eb-9cb5-b520a0217c62.png)
4. To teach recipes, you can use the (pre-existing) plus button or (new) drag ingredients directly from the JEI GUI (without needing cheat mode):
![image](https://user-images.githubusercontent.com/539951/119514528-33cbc300-bdc9-11eb-8d1b-129b384f1fbe.png)

uecasm committed:
Move tag ids to API.
(Strictly speaking not needed any more, but it still seems like a good idea.)

uecasm committed:
Add client-side view and JEI GUI handlers.

uecasm committed:
Initial crafting modules, plus JEI integration.

uecasm committed:
Fix custom recipe loading on dedicated server.
Also sort recipes into some semblance of useful order.

uecasm commented:
I'm still considering a redesign on the crafting module structure -- rather than having a single interface with policy subclasses, it might be better to have the base `ICraftingBuildingModule` (used directly by custom-only modules), and extend `ICraftingCraftingBuildingModule` and `ISmeltingCraftingBuildingModule` (for each vanilla crafting type) from that.  (Or with better names.)  This would let them have different interfaces, which might be useful once state starts to be added.

On the other hand, the present design is simpler and more easily allows a single ModuleView to serve all cases.

Hard to say which one will be better in the end, it depends how similar the remaining parts of implementation end up being.  Still, it could always be changed later if needed.

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java:
```diff
@@ -314,9 +311,36 @@ public void switchIndex(final int i, final int j)
+     */
+    protected boolean isRecipeCompatibleWithCraftingModule(final IToken<?> token)
+    {
+        return this.getModules(ICraftingBuildingModule.class).stream()
```
I have a feeling that a simple for each loop here could be much easier and much more efficient.

uecasm committed:
Just use a loop.

Also, move the intermediate check somewhere that a module can override it, in case we want to support more exotic recipes.

uecasm committed:
Merge remote-tracking branch 'remotes/upstream/version/main' into jei-crafting2

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/smelter/EntityAIWorkSmelter.java:
```diff
@@ -399,15 +406,76 @@ private int getRequiredProgressForMakingRawMaterial()
     }
+
+    @NotNull
+    public static List<IGenericRecipe> getAdditionalRecipesForDisplayPurposesOnly()
```
I'm really unsure about all of this, basically, this is the #1 feature that I'm unsure about and wondering about scrapping/modifying it. That's why I'm unsure if I want people to actually see this so obvious in JEI atm.

uecasm replied:
If the feature itself is removed, the JEI display can be removed at the same time (that's an advantage of having them both in the same place rather than in entirely different files).

While it exists, though, it ought to be documented in some way, otherwise nobody would know that it exists.

The other option might be to remove the JEI/crafting module entirely -- which is not unreasonable since the smelter doesn't support crafterrecipes or teachable recipes at all, but it would make the mod less self-documenting.  (I was considering later adding a JEI page for the animal huts too, showing the animal drops as outputs plus some special cases like cow milking.  This would use the same mechanism, although is a bit more complicated due to loot tables.)

someaddons replied:
there is some stuff which is intentionally not shown/displayed as readily available public information such as wiki/jei is

uecasm replied:
Sure, and that makes sense for some things, like specific formulas for raids and XP etc.  It makes less sense to hide crafting mechanics, though.

uecasm committed:
Remove tool smelting (#7178)

uecasm committed:
Merge remote-tracking branch 'remotes/upstream/version/main' into jei-crafting2

uecasm committed:
Weak `this` was pointless due to lambda capture.

uecasm committed:
Merge remote-tracking branch 'remotes/upstream/version/main' into jei-crafting2

uecasm committed:
Merge branch 'version/main' into recipe/jei-crafting

marchermans commented:
:ship it:

marchermans commented:
:shipit:

uecasm committed:
Merge branch 'version/main' into recipe/jei-crafting

uecasm committed:
Merge remote-tracking branch 'remotes/upstream/version/main' into jei-crafting2
